unlock-app/src/hooks/useMultipleRecipient.ts

```diff
@@ -0,0 +1,193 @@
+import { useState, useEffect, useContext } from 'react'
+import { ToastHelper } from '../components/helpers/toast.helper'
+import { PaywallConfig } from '../unlockTypes'
+import { ConfigContext } from '../utils/withConfig'
+import { WalletServiceContext } from '../utils/withWalletService'
+import { getAddressForName } from './useEns'
+import { purchaseMultipleKeys } from './useLock'
+
+interface User {
+  lockAddress: string
+  userAddress: string
+}
+export interface RecipientItem {
+  userAddress: string
+  resolvedAddress: string
+  valid: boolean
+  index: number
+  metadata?: { [key: string]: any }
+  keyId?: string
+}
+
+interface RecipientPayload {
+  users: {
+    userAddress: string
+    metadata: { [key: string]: any }
+    lockAddress: string
+  }[]
+}
+
+export const useMultipleRecipient = (
+  account?: string,
+  paywallConfig?: PaywallConfig,
+  lockAddress?: string
+) => {
+  const walletService = useContext(WalletServiceContext)
+  const [hasMultipleRecipients, setHasMultipleRecipients] = useState(false)
+  const [recipients, setRecipients] = useState(new Set<RecipientItem>())
+  const [loading, setLoading] = useState(false)
+  const config: any = useContext(ConfigContext)
+  const [retryBulkAction, setRetryBulkAction] = useState(true)
+  const { maxRecipients = 1, locks } = paywallConfig ?? {}
+  const lock = lockAddress ? locks?.[lockAddress] : {}
+
+  const normalizeRecipients = (ignoreRecipients?: User[]) => {
+    if (!lockAddress) return
+
+    const listWithoutExcluded = recipientsList().filter(
+      ({ resolvedAddress }) => {
+        return ignoreRecipients && ignoreRecipients?.length > 0
+          ? !ignoreRecipients
+              .map(({ userAddress }) => userAddress)
+              .includes(resolvedAddress)
+          : true
+      }
+    )
```

Comment:
Can you refactor this and add some comments? It's hard to understand.

Reply:
@searchableguy done! hope more readable now!